numpy/core/src/umath/ufunc_type_resolution.c

```diff
@@ -1508,6 +1507,22 @@ ufunc_loop_matches(PyUFuncObject *self,
             tmp = dtypes[i];
             Py_INCREF(tmp);
         }
+        else if (types[i] == PyArray_TYPE(op[i])) {
+            /*
+             * Don't bother doing casting checks if the type numbers exactly
+             * match.
```

Comment:
My concern with a ufunc taking multiple variables is that we don't have any way to distinguish between ufuncs that accept arguments with the same time units or any time units. E.g., if `S` and `T` are typevar variables, distinguishing between signatures `(datetime64[T], datetime64[T])` and `(datetime64[S], datetime64[T])`.

In practice, I'm not sure this is a major concern for ufuncs added in NumPy, because unit tests should catch these sorts of issues. But at least in theory, datetime64 ufuncs can also be written in third party code (e.g., [with Numba](https://github.com/numba/numba/issues/917#issuecomment-69011988)).

Basically, I think it's important to emphasize that this is a special case rule in the comments and limit it's scope as much as possible to ensure as few complications as possible.